File: Ejercicios/T7/PYTHON/Adestra_detector_peons.py
import os
import glob
import cv2
import numpy as np


# returns os path as imaxe dun cartafol
# coas extensions definidas en imgExts
def getImagePaths(folder, imgExts):
  imagePaths = []
  for x in os.listdir(folder):
    xPath = os.path.join(folder, x)
    if os.path.splitext(xPath)[1] in imgExts:
      imagePaths.append(xPath)
  return imagePaths


# Lemos as imaxes dun cartafol
# return a lista de imaxe e etiquetas
def getDataset(folder, classLabel):
  images = []
  labels = []
  imagePaths = getImagePaths(folder, ['.jpg', '.png', '.jpeg'])
  for imagePath in imagePaths:
    # print(imagePath)
    im = cv2.imread(imagePath, cv2.IMREAD_COLOR)
    images.append(im)
    labels.append(classLabel)
  return images, labels

# achas as caracteristicas HOG
def computeHOG(hog, images):
  hogFeatures = []
  for image in images:
    hogFeature = hog.compute(image)
    hogFeatures.append(hogFeature)
  return hogFeatures


# cambiamos formato de HoG para SVM
def prepareData(hogFeatures):
  featureVectorLength = len(hogFeatures[0])
  data = np.float32(hogFeatures).reshape(-1, featureVectorLength)
  return data

#Inicializa SVM 
def svmInit(C, gamma):
  model = cv2.ml.SVM_create()
  model.setGamma(gamma)
  model.setC(C)
  model.setKernel(cv2.ml.SVM_LINEAR)
  model.setType(cv2.ml.SVM_C_SVC)
  model.setTermCriteria((cv2.TERM_CRITERIA_EPS +
                         cv2.TERM_CRITERIA_MAX_ITER,
                         1000, 1e-3))
  return model


# adestramos o SVM sobre os datos e as etiquetas
def svmTrain(model, samples, labels):
  model.train(samples, cv2.ml.ROW_SAMPLE, labels)


# predecimos as etiquetas
def svmPredict(model, samples):
  return model.predict(samples)[1]


# Avaliamos o modelos
# tendo en conta as predicion e o ground truth
def svmEvaluate(model, samples, labels):
  labels = labels[:, np.newaxis]
  pred = model.predict(samples)[1]
  correct = np.sum((labels == pred))
  err = (labels != pred).mean()
  print('label -- 1:{}, -1:{}'.format(np.sum(pred == 1),
          np.sum(pred == -1)))
  return correct, err * 100


# creamos o directorio
def createDir(folder):
  try:
    os.makedirs(folder)
  except OSError:
    print('{}: xa existe'.format(folder))
  except Exception as e:
    print(e)

#Inicializa HOG 
winSize = (64, 128)
blockSize = (16, 16)
blockStride = (8, 8)
cellSize = (8, 8)
nbins = 9
derivAperture = 1
winSigma = -1
histogramNormType = 0
L2HysThreshold = 0.2
gammaCorrection = True
nlevels = 64
signedGradient = False

# Inicializa HOG
hog = cv2.HOGDescriptor(winSize, blockSize, blockStride,
                      cellSize, nbins,derivAperture,
                      winSigma, histogramNormType, L2HysThreshold,
                      gammaCorrection, nlevels,signedGradient)

# Flags para seleccionar adestarmento, test e consulta
trainModel = True
testModel = True
queryModel = True

# Para a base de datos do INRIA de personas
rootDir = '../data/images/INRIAPerson/'

# Path para cartafoles de adestramento e test
trainDir = os.path.join(rootDir, 'train_64x128_H96')
testDir = os.path.join(rootDir, 'test_64x128_H96')

# ================================ adestramento =====================
if trainModel:
    # Lemos imaxe pos e neg
    trainPosDir = os.path.join(trainDir, 'posPatches')
    trainNegDir = os.path.join(trainDir, 'negPatches')

    # etiqueta 1 para os positios e -1 para os negativos
    trainPosImages, trainPosLabels = getDataset(trainPosDir, 1)
    trainNegImages, trainNegLabels = getDataset(trainNegDir, -1)

    # comprobamos tamanhos
    print(set([x.shape for x in trainPosImages]))
    print(set([x.shape for x in trainNegImages]))

    # informamos do que lemos no disco
    print('positivo - {}, {} || negativo - {}, {}'
        .format(len(trainPosImages),len(trainPosLabels),
        len(trainNegImages),len(trainNegLabels)))

    # Concatenamos imaxes/etiqueta para positivo/negativo
    trainImages = np.concatenate((np.array(trainPosImages),
                       np.array(trainNegImages)),
                                  axis=0)
    trainLabels = np.concatenate((np.array(trainPosLabels),
                       np.array(trainNegLabels)),
                                  axis=0)
    # Hog 
    hogTrain = computeHOG(hog, trainImages)

    # formato para  SVM
    trainData = prepareData(hogTrain)

    # Comprobamos dimensions
    print('trainData: {}, trainLabels:{}'
            .format(trainData.shape, trainLabels.shape))
    # Creamos obxecto SVM e entrenamos
    model = svmInit(C=0.01, gamma=0)
    svmTrain(model, trainData, trainLabels)
    model.save('../data/models/pedestrian.yml')

# ================================ Test  ===============
if testModel:
    # cargamos dende disco ficheir xml
    model = cv2.ml.SVM_load('../data/models/pedestrian.yml')
    # avaliamos o modelo
    testPosDir = os.path.join(testDir, 'posPatches')
    testNegDir = os.path.join(testDir, 'negPatches')
    
    testPosImages, testPosLabels = getDataset(testPosDir, 1)
    testNegImages, testNegLabels = getDataset(testNegDir, -1)

    # HoG sobre as imaxes
    hogPosTest = computeHOG(hog, np.array(testPosImages))
    testPosData = prepareData(hogPosTest)
    
    # Clasificamos e calcualos TP e FP 
    posCorrect, posError = svmEvaluate(model, testPosData, 
                                       np.array(testPosLabels))

    tp = posCorrect
    fp = len(testPosLabels) - posCorrect
    print('TP: {}, FP: {}, Total: {}, error: {}'
            .format(tp, fp, len(testPosLabels), posError))
    # O mesmo pero para imaxes negativas
    hogNegTest = computeHOG(hog, np.array(testNegImages))
    testNegData = prepareData(hogNegTest)
    negCorrect, negError = svmEvaluate(model, testNegData, 
                                       np.array(testNegLabels))

    # achamos TP e FP
    tn = negCorrect
    fn = len(testNegData) - negCorrect
    print('TN: {}, FN: {}, Total: {}, error: {}'
            .format(tn, fn, len(testNegLabels), negError))
    # achamos Precision e Recall
    precision = tp * 100 / (tp + fp)
    recall = tp * 100 / (tp + fn)
    print('Precision: {}, Recall: {}'.format(precision, recall))

# ================================ Consulta =============================================
# Executar o detector de obxectos nunha imaxe de consulta para atopar peóns
# Cargaremos o modelo de novo e probaremos o modelo
# Isto é só para explicar como cargar un modelo SVM
# Podes usar o modelo directamente dende memoria (xa o temos aí)
if queryModel:
    # cargamos o modelo
    model = cv2.ml.SVM_load('../data/models/pedestrian.yml')
    sv = model.getSupportVectors()
    rho, aplha, svidx = model.getDecisionFunction(0)
    svmDetector = np.zeros(sv.shape[1] + 1, dtype=sv.dtype)
    svmDetector[:-1] = -sv[:]
    svmDetector[-1] = rho

  
    hog.setSVMDetector(svmDetector)

    filename = "../data/images/pedestrians/3.jpg"
    queryImage = cv2.imread(filename, cv2.IMREAD_COLOR)

    #executamos o noso detector nunha imaxe cunha altura fixa
    finalHeight = 800.0
    scale = finalHeight / queryImage.shape[0]
    queryImage = cv2.resize(queryImage, None, fx=scale, fy=scale)

    # detectMultiScale detectará a n niveis de imaxe aumentando a escala
     # e reducindo o tamaño da imaxe por un factor de 1,05
    bboxes, weights = hog.detectMultiScale(queryImage, winStride=(8, 8),
                                           padding=(32, 32), scale=1.05,
                                           finalThreshold=2, hitThreshold=1.0)
    # Visualizamos as caixas na iamxe
    for bbox in bboxes:
      x1, y1, w, h = bbox
      x2, y2 = x1 + w, y1 + h
      cv2.rectangle(queryImage, (x1, y1), (x2, y2),
                    (0, 255, 0), thickness=3, lineType=cv2.LINE_AA)
    cv2.imshow("Imaxe consulta",queryImage)
    cv2.waitKey(0)

# Iniciamos HOG
winSize = (64, 128)
blockSize = (16, 16)
blockStride = (8, 8)
cellSize = (8, 8)
nbins = 9
derivAperture = 1
winSigma = -1
histogramNormType = 0
L2HysThreshold = 0.2
gammaCorrection = True
nlevels = 64
signedGradient = False


hog = cv2.HOGDescriptor(winSize, blockSize, blockStride,
                        cellSize, nbins, derivAperture,
                        winSigma, histogramNormType, L2HysThreshold,
                        gammaCorrection, nlevels, signedGradient)

# Cargamos o modelo entrenao por nos
model = cv2.ml.SVM_load('../data/models/pedestrian.yml')
sv = model.getSupportVectors()
rho, aplha, svidx = model.getDecisionFunction(0)
svmDetectorTrained = np.zeros(sv.shape[1] + 1, dtype=sv.dtype)
svmDetectorTrained[:-1] = -sv[:]
svmDetectorTrained[-1] = rho
# iniciamos SVMDetector entrenado por nos en HOG
hog.setSVMDetector(svmDetectorTrained)

# Detector de persoas de OpenCV con HoG
hogDefault = cv2.HOGDescriptor(winSize, blockSize, blockStride,
                               cellSize, nbins, derivAperture,
                               winSigma, histogramNormType,
                               L2HysThreshold,gammaCorrection,
                               nlevels, signedGradient)
svmDetectorDefault = cv2.HOGDescriptor_getDefaultPeopleDetector()
hogDefault.setSVMDetector(svmDetectorDefault)

# lemos as imaxes do directorio de persoas
imageDir = '../data/images/pedestrians'
imagePaths = glob.glob(os.path.join(imageDir, '*.jpg'))

#executamos o detector para imaxes con altura fixa
finalHeight = 800.0

for imagePath in imagePaths:
    print('procesando: {}'.format(imagePath))

     # Lemos a imaxe
    im = cv2.imread(imagePath, cv2.IMREAD_COLOR)

    # redimensionamos a imaxe para que teña unha altura prederminada de finalHeight
    scale = finalHeight / im.shape[0]
    im = cv2.resize(im, None, fx=scale, fy=scale)

      # Detecta persoas usando detectores SVM adestrados e predeterminados
    # detectMultiScale usando un detector adestrado por nós
    bboxes, weights = hog.detectMultiScale(im, winStride=(8, 8),
                                    padding=(32, 32),scale=1.05,
                                    finalThreshold=2,hitThreshold=1.0)

   # detectMultiScale modelos predetemiando
    bboxes2, weights2 = hogDefault.detectMultiScale(im, winStride=(8, 8),
                                padding=(32, 32),scale=1.05,
                                finalThreshold=2,hitThreshold=0)

# Debuxa rectángulos atopados na imaxe. Debuxaremos
     # caixas verdes para persoas detectadas polo modelo adestrado e
     # caixas vermellas para as persoas detectadas polo modelo predeterminado de OpenCV.
    if len(bboxes) > 0:
        print('Detector entrenado :: peons detectados: {}'
                .format(bboxes.shape[0]))
    if len(bboxes2) > 0:
        print('Detector de OpCV :: peons detectados: {}'
                .format(bboxes2.shape[0]))

    # Debuxa caixas delimitadoras detectadas sobre a imaxe
    # Vermello = detector predeterminado, Verde = Detector adestrado por nos
    for bbox in bboxes:
        x1, y1, w, h = bbox
        x2, y2 = x1 + w, y1 + h
        cv2.rectangle(im, (x1, y1), (x2, y2),
                      (0, 255, 0), thickness=3,
                      lineType=cv2.LINE_AA)

    for bbox in bboxes2:
        x1, y1, w, h = bbox
        x2, y2 = x1 + w, y1 + h
        cv2.rectangle(im, (x1, y1), (x2, y2),
                      (0, 0, 255), thickness=3,
                      lineType=cv2.LINE_AA)
    # Visualizamos os resultados finais
    cv2.imshow("Imaxe",im)
    # salvamos a disco
    imResultPath = os.path.join('resultados', os.path.basename(imagePath))
    cv2.imwrite(imResultPath, im)
    cv2.waitKey(0)

File: Ejercicios/T7/C++/Adestra_detector_peons.cpp


#include <iostream>
#include <dirent.h>
#include <opencv2/highgui.hpp>
#include <opencv2/imgproc.hpp>
#include "opencv2/objdetect.hpp"
#include <opencv2/ml.hpp>

using namespace cv::ml;
using namespace cv;
using namespace std;

// returns os path as imaxe dun cartafol definidas en imgExts
void getFileNames(string dirName, vector<string> &imageFnames)
{
  DIR *dir;
  struct dirent *ent;
  int count = 0;
  string imgExt = "jpg";

  vector<string> files;
  if ((dir = opendir (dirName.c_str())) != NULL) {
    while ((ent = readdir (dir)) != NULL) {
      //evitar ler nomes non desexados
      if(strcmp(ent->d_name,".") == 0 | strcmp(ent->d_name, "..") == 0) { continue; }
      string temp_name = ent->d_name;
      files.push_back(temp_name);
    }

    // Ordeamos
    std::sort(files.begin(),files.end());
    for(int it=0;it<files.size();it++) {
      string path = dirName;
      string fname = files[it];
      if (fname.find(imgExt, (fname.length() - imgExt.length())) != std::string::npos) {
        path.append(fname);
        imageFnames.push_back(path);
      }
    }
    closedir (dir);
  }
}

// Lemos imaxes e etiquetas
void getDataset(string &pathName, int classVal, vector<Mat> &images, vector<int> &labels) {
  vector<string> imageFiles;
  getFileNames(pathName, imageFiles);
  for (int i = 0; i < imageFiles.size(); i++) {
    Mat im = imread(imageFiles[i]);
    images.push_back(im);
    labels.push_back(classVal);
  }
}

// Inicializa HOG
HOGDescriptor hog(
        Size(64, 128), //winSize
        Size(16, 16),  //blocksize
        Size(8, 8),    //blockStride,
        Size(8, 8),    //cellSize,
                9,     //nbins,
                0,     //derivAperture,
                -1,    //winSigma,
                HOGDescriptor::HistogramNormType (0),     //histogramNormType,
                0.2,   //L2HysThresh,
                1,     //gammal correction,
                64,    //nlevels=64
                0);    //signedGradient

// acha HOG
void computeHOG(vector<vector<float> > &hogFeatures, vector<Mat> &images)
{
  for(int y = 0; y < images.size(); y++)
  {
    vector<float> descriptor;
    hog.compute(images[y], descriptor);
    hogFeatures.push_back(descriptor);
  }
}

// formato para SVM
void prepareData(vector<vector<float> > &hogFeatures, Mat &data)
{
  int descriptorSize = hogFeatures[0].size();

  for(int i = 0; i < hogFeatures.size(); i++)
    for(int j = 0; j < descriptorSize; j++)
      data.at<float>(i,j) = hogFeatures[i][j];
}

// Iniciamos SVM
Ptr<SVM> svmInit(float C, float gamma)
{
  Ptr<SVM> svm = SVM::create();
  svm->setGamma(gamma);
  svm->setC(C);
  svm->setKernel(SVM::LINEAR);
  svm->setType(SVM::C_SVC);
  return svm;
}

// Adestramos SVM
void svmTrain(Ptr<SVM> svm, Mat &trainMat, vector<int> &trainLabels)
{
  Ptr<TrainData> td = TrainData::create(trainMat, ROW_SAMPLE, trainLabels);
  svm->train(td);
}

// predecimos labels
void svmPredict(Ptr<SVM> svm, Mat &testMat, Mat &testResponse)
{
  svm->predict(testMat, testResponse);
}

// avaliamos o modelo
void svmEvaluate(Mat &testResponse, vector<int> &testLabels, int &correct, float &error)
{
  for(int i = 0; i < testResponse.rows; i++)
  {
    if(testResponse.at<float>(i,0) == testLabels[i])
      correct = correct + 1;
  }
  error = (testLabels.size() - correct)*100.0/testLabels.size();
}

int main()
{
  // Flags para controlar o comportamento do programa
  bool trainModel = 1;
  bool testModel = 1;

  // Path ao dataset do INRIA
  string rootDir = "../../data/images/INRIAPerson/";

  // Paths aos directorios de adestramento e test
  string trainDir = rootDir + "train_64x128_H96/";
  string testDir = rootDir + "test_64x128_H96/";

  // ================================ Adestramento =============================================
  if (trainModel == 1) {
    string trainPosDir = trainDir + "posPatches/";
    string trainNegDir = trainDir + "negPatches/";

    vector<Mat> trainPosImages, trainNegImages;
    vector<int> trainPosLabels, trainNegLabels;

    //  1 positivos e -1 para imaxes negativas
    getDataset(trainPosDir, 1, trainPosImages, trainPosLabels);
    getDataset(trainNegDir, -1, trainNegImages, trainNegLabels);


    cout << "positivas - " << trainPosImages.size() << " , " << trainPosLabels.size() << endl;
    cout << "negativas - " << trainNegImages.size() << " , " << trainNegLabels.size() << endl;


    vector<Mat> trainImages;
    vector<int> trainLabels;
    trainImages = trainPosImages;
    trainImages.insert(trainImages.end(), trainNegImages.begin(), trainNegImages.end());

    trainLabels = trainPosLabels;
    trainLabels.insert(trainLabels.end(), trainNegLabels.begin(), trainNegLabels.end());

    // Acha HOG
    vector<vector<float> > hogTrain;
    computeHOG(hogTrain, trainImages);

    // formato SVM
    int descriptorSize = hogTrain[0].size();
    cout << "Descriptor Size : " << descriptorSize << endl;
    Mat trainData(hogTrain.size(), descriptorSize, CV_32FC1);
    prepareData(hogTrain, trainData);

    // Inicializa SVM
    float C = 0.01, gamma = 0;
    Ptr<SVM> svm = svmInit(C, gamma);
    svmTrain(svm, trainData, trainLabels);
    svm->save("../results/pedestrian.yml");
  }

  // ================================ Test =============================================
  if (testModel == 1) {
    // cargamos o modelo dende disco
    Ptr<SVM> svm = ml::SVM::load("../results/pedestrian.yml");

    string testPosDir = testDir + "posPatches/";
    string testNegDir = testDir + "negPatches/";

    vector<Mat> testPosImages, testNegImages;
    vector<int> testPosLabels, testNegLabels;
    getDataset(testPosDir, 1, testPosImages, testPosLabels);
    getDataset(testNegDir, -1, testNegImages, testNegLabels);

    cout << "positivos - " << testPosImages.size() << " , " << testPosLabels.size() << endl;
    cout << "negativos - " << testNegImages.size() << " , " << testNegLabels.size() << endl;

    // =========== Test sobre imaxes positivas ===============
    // Computa HOG
    vector<vector<float> > hogPosTest;
    computeHOG(hogPosTest, testPosImages);

    // formato para SVM
    int descriptorSize = hogPosTest[0].size();
    cout << "Tamanho do descriptor : " << descriptorSize << endl;
    Mat testPosData(hogPosTest.size(), descriptorSize, CV_32FC1);
    prepareData(hogPosTest, testPosData);
    cout << testPosData.rows << " " << testPosData.cols << endl;

    // clasificamos
    Mat testPosPredict;
    svmPredict(svm, testPosData, testPosPredict);
    int posCorrect = 0;
    float posError = 0;
    svmEvaluate(testPosPredict, testPosLabels, posCorrect, posError);

    // achamos TP FP
    int tp = posCorrect;
    int fp = testPosLabels.size() - posCorrect;
    cout << "TP: " << tp << " FP: " << fp << " total: " << testPosLabels.size() << " error: " << posError << endl;

    // =========== Test sobre imaxe negativas ===============
    // HOG
    vector<vector<float> > hogNegTest;
    computeHOG(hogNegTest, testNegImages);

    // formato para SVM
    cout << "Tamanho do descriptor : " << descriptorSize << endl;
    Mat testNegData(hogNegTest.size(), descriptorSize, CV_32FC1);
    prepareData(hogNegTest, testNegData);

    // calsificamos
    Mat testNegPredict;
    svmPredict(svm, testNegData, testNegPredict);
    int negCorrect = 0;
    float negError = 0;
    svmEvaluate(testNegPredict, testNegLabels, negCorrect, negError);

    // achamos TP e FP
    int tn = negCorrect;
    int fn = testNegLabels.size() - negCorrect;
    cout << "TN: " << tn << " FN: " << fn << " total: " << testNegLabels.size() << " error: " << negError << endl;

    // achamos Precision e Recall
    float precision = tp * 100.0 / (tp + fp);
    float recall = tp * 100.0 / (tp + fn);
    cout << "Precision: " << precision << " Recall: " << recall << endl;
  }
  return 0;
}
